### PR TITLE
Reporter backends plugin hook

### PR DESCRIPTION
**Pull request by travishathaway**
### Description

This pull request completes the first phase of the reporter backends plugin hook roll out. It allows plugin authors to customize the following aspects of conda's terminal output:

- Progress bars
- Spinners
- Confirmation inputs
- `conda info` layout

Please see this epic for more information:

- https://github.com/conda/conda/issues/13707

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](https://github.com/conda/conda/blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: docs/source/dev-guide/plugins/reporter_backends.rst [hunk 1]
```diff
@@ -0,0 +1,162 @@
+=================
+Reporter Backends
+=================
+
+A reporter backend is a plugin hook that allows you to change the output displayed
+for existing conda commands. Internally, conda uses this to differentiate between
+displaying normal, "console" output or displaying "json" output. An example of each
+of these reporter backend types is shown below:
```
Comment (beeankha) [inline docs/source/dev-guide/plugins/reporter_backends.rst:8]: Should there be a couple of example code blocks after this paragraph?

File: docs/source/dev-guide/plugins/reporter_backends.rst [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    from conda import plugins
+    from conda.plugins.types import (
+        CondaReporterBackend,
+        ReporterRendererBase,
+        ProgressBarBase,
+    )
+
+
+    class PprintReporterRenderer(ReporterRendererBase):
+        """ "
```
Comment (beeankha) [inline docs/source/dev-guide/plugins/reporter_backends.rst:65]:
```suggestion
        """
```

**Review by beeankha — DISMISSED** [review]

File: docs/source/dev-guide/plugins/reporter_backends.rst (hunk 1 above)
Comment (travishathaway) [reply docs/source/dev-guide/plugins/reporter_backends.rst:8]: Thanks for catching this. This paragraph shouldn't actually be there and was something I forgot to update when last working with the code. I will replace it with something more suitable.

**Review by beeankha — DISMISSED** [review]

File: conda/cli/main_search.py [hunk 3]
```diff
@@ -231,11 +227,9 @@ def execute(args: Namespace, parser: ArgumentParser) -> int:
...
         return 0
 
-    with Spinner(
-        "Loading channels",
-        not context.verbose and not context.quiet,
-        context.json,
-    ):
+    spinner = get_spinner("Loading channels")
+
+    with spinner:
```
Comment (jezdez) [inline conda/cli/main_search.py:232]:
```suggestion
    with get_spinner("Loading channels"):
```

File: conda/core/package_cache_data.py [hunk 4]
```diff
@@ -782,143 +785,156 @@ def execute(self):
...
+                    future.add_done_callback(
                         partial(
                             done_callback,
-                            actions=(cache_action, extract_action),
+                            actions=(cache_action,),
                             exceptions=exceptions,
-                            progress_bar=progress_bars[prec_or_spec],
-                            finish=True,
+                            progress_bar=progress_bar,
+                            finish=False,
```
Comment (jezdez) [inline conda/core/package_cache_data.py:858]: That change of the default to `finish`, why is this?

File: conda/core/package_cache_data.py [hunk 5]
```diff
@@ -782,143 +785,156 @@ def execute(self):
...
+                if IS_INTERACTIVE:
+                    print("\r")  # move to column 0
+                else:
+                    print(" done")
+
+            if exceptions:
+                # avoid printing one CancelledError() per pending download
+                not_cancelled = [
+                    e for e in exceptions if not isinstance(e, CancelledError)
+                ]
```
Comment (jezdez) [inline conda/core/package_cache_data.py:911]:
```suggestion
                not_cancelled = [
                    exception
                    for exception in exceptions
                    if not isinstance(exception, CancelledError)
                ]
```

File: news/14206-spinner-support-reporter-backends [hunk 6]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Adds support for defining spinners for the reporter backends plugin hook (#14206)
```
Comment (jezdez) [inline news/14206-spinner-support-reporter-backends:3]:
```suggestion
* Adds support for defining spinners for the reporter backends plugin hook. (#14206)
```

File: news/13868-reporter-backend-and-reporter-output-plugin-hooks [hunk 7]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* Adds two new plugin hooks for reporter backends and reporter outputs (#13868)
```
Comment (jezdez) [inline news/13868-reporter-backend-and-reporter-output-plugin-hooks:3]:
```suggestion
* Adds a new plugin hook for reporter backends. (#13868)
```

File: conda/plugins/reporter_backends/json.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+@hookimpl(
+    tryfirst=True
+)  # make sure the default json reporter backend can't be overridden
+def conda_reporter_backends():
+    """
+    Reporter backend for JSON
+
+    This is the default reporter backend that returns objects as JSON strings
+    that can be passed to reporter outputs.
```
Comment (jezdez) [inline conda/plugins/reporter_backends/json.py:121]:
```suggestion
    This is the default reporter backend that returns objects as JSON strings.
```

**Review by jezdez — COMMENTED** [review]
Minor nits, I think this is good to go though

File: conda/cli/main_info.py [hunk 9]
```diff
@@ -390,80 +386,176 @@ def builder():
...
+
+        output = [
+            f"sys.version: {sys.version[:40]}...",
+            f"sys.prefix: {sys.prefix}",
+            f"sys.executable: {sys.executable}",
+            "conda location: {}".format(self._info_dict["conda_location"]),
+        ]
+
+        for cmd in sorted(set(find_commands() + ("build",))):
+            output.append("conda-{}: {}".format(cmd, find_executable("conda-" + cmd)))
```
[the remark this answers is not in the document]
Comment (travishathaway) [reply conda/cli/main_info.py:480]: Not sure I follow. We are constructing a string here. We are not printing anything.

Can you please clarify your concern a little more?

File: conda/core/package_cache_data.py (hunk 4 above)
Comment (travishathaway) [reply conda/core/package_cache_data.py:858]: To be honest, I cannot remember why that was done! I can revert it back though as I do not see any downsides.

**Comment by beeankha** [comment]
pre-commit.ci autofix

**Review by beeankha — APPROVED** [review]

**Review by jezdez — APPROVED** [review]

**Pull request merged**